examples/gno.land/p/grc/grc20/impl/impl.gno

```diff
@@ -51,22 +52,32 @@ func (t *Token) GetDecimals() uint   { return t.decimals }
 func (t *Token) TotalSupply() uint64 { return t.totalSupply }
 
 func (t *Token) BalanceOf(address std.Address) uint64 {
+	address = users.GetAddressByInput(address.String())
```

Comment:
Or maybe to distinguish, TransferByInput.  But I kinda think the caller should resolve to addr first.

Reply:
That’s where I would be more « gno strict » than you? :)

Yes, this is the way for a generic library,  and this is how I used to to it first; it matches with library that can be used in any context.

In our case, I find it appropriate to enforce this kind of features:
* first, as an example of how connected is Gno (one of the issues of ENS today is that it’s not connected enough; gno could be fully connected with simplicity
* two, the only drawback for me with this method is that we should be sure that it does not add any security issue, but, I consider that if we have this fear, we should not limit ourselves, but just be sure that there is no flaw (support edge cases, write more tests etc)
* and third, this package is the official implementation, it’s still possible to fork it and just follow the interface

So yes, I’m opinionated in the context of gno to encourage inter-realms communication in the « official(/first) packages »; where in real Go life, I would do the opposite

What do you think?

Reply:
I’ve made the suggested renaming from the previous comment

Waiting for your feedback to either keep the opinionated forced-import in this implementation or make as you suggested initially, and move this logic in r/foo20

Reply:
I just pushed a new commit that keeps my opinionated preference but where I go more in what you suggest in the first comment `nameOrAddress string` instead of `address std.Address`

But since most of the methods are not taking `address`, but `owner`, or worst `owner, spender, to`; I preferred not to specify `ownerNameOrAddress`, `spenderNameOrAddress`, etc

And btw, by thinking more about it, if you don’t like to enforce my opinionated approach here, I can split the implementation into two ones, a `simpleimpl` which is basically like before this PR; and an `impl` that just adds the resolve automatically; but since we have the interface available on the upper-package; I don’t think that we need to push this case and would encourage having less packages, and less method choices in general

I consider that either you’re a top-level contract developer and want to use the opinionated/suggested approaches, either you’re a low-level contract developer and won’t care about rewriting things in the way you want

Reply:
I would leave all the original std.Address in most function signatures, like `func (t *Token) BalanceOf(address std.Address) uint64 {` rather than address string. For example, if someone creates a username like "g1..." then this would pose a problem, and I like that the type-checking system checks for this.

Reply:
what do you think about:

1. adding a new `type users.NameOrAddress string`
2. improve security by not resolving a username that looks like an address
3. update this PR to use `users.NameOrAddress` everywhere?

Reply:
thank you for #166, I'll continue with that base :+1: